internal/pkg/service/objectstorage/cloud/azure/storage.go

```diff
@@ -209,18 +210,51 @@ func (s AzureObjectStorageAdapter) CreateBucket(ctx context.Context, bucket *v1a
 // We want to prevent the Storage Account from being used by anyone
 func (s AzureObjectStorageAdapter) DeleteBucket(ctx context.Context, bucket *v1alpha1.Bucket) error {
 	storageAccountName := s.getStorageAccountName(bucket.Spec.Name)
-	// We delete the Storage Account, which delete the Storage Container
-	_, err := s.storageAccountClient.Delete(
+	// Check if Storage Account exists on Azure
+	existsStorageAccount, err := s.existsStorageAccount(ctx, storageAccountName)
+	if err != nil {
+		return err
+	}
+	// If Storage Account does not exists, we need to create it first
+	if !existsStorageAccount {
+		return nil
+	}
+
+	// Check if the Storage Container is empty before proceeding to delete it
+	containerResponse, err := s.blobContainerClient.Get(ctx, s.cluster.GetResourceGroup(), storageAccountName,
+		bucket.Spec.Name, &armstorage.BlobContainersClientGetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the response is correctly configured to avoid panics
+	if containerResponse.ContainerProperties == nil ||
+		containerResponse.ContainerProperties.Metadata == nil ||
+		containerResponse.ContainerProperties.Metadata["ItemCount"] == nil {
+		return fmt.Errorf("ItemCount metadata not found for storage container %s", bucket.Spec.Name)
+	}
+
+	// Get the number of items in the container
+	itemCount, err := strconv.Atoi(*containerResponse.ContainerProperties.Metadata["ItemCount"])
+	if err != nil {
+		return fmt.Errorf("conversion failed for storage container %s", bucket.Spec.Name)
+	}
+	if itemCount > 0 {
+		return fmt.Errorf("storage container %s is not empty", bucket.Spec.Name)
+	}
```

Comment:
Why do we add this check here ?

Reply:
Where would you put it otherwise ?

Reply:
We want to prevent délétion if thé container is not empty

Reply:
Well that's the thing, no we do not want the container to be removed because then we lose all data for things like loki and mimir for instance.

We are adding this to make sure the deletion is intentionnal and not an accident as is explained here https://github.com/giantswarm/giantswarm/issues/31006

We created this issue in the first place because an upgrade of the loki chart broke the extraObjects section that we use for bucket creations as long as there was more than 1 object in the extraObject sections which we had on glippy, so we lost all existing buckets for our onprem test installations.

This could also have happened when we enabled the loki ruler (if we did not see this before) as we will use another bucket and we would have lost all data from all installations

Reply:
The main reason we did not have this check earlier is because it was there by default on aws and we assumed it was the same on azure I guess

Reply:
Thanks for explaining